### PR TITLE
Policy fix revert by only recording an "old" value if it existed before the changes

### PR DESCRIPTION
**Pull request by jrajahalme**
Only record an old entry in `ChangeState` if it existed before this round of changes. We do this by testing if the entry is already in `Adds`. If not, then we record the old entry key and value. If the `Adds` entry exists, however, this entry may have only been added on this round of changes and we do not record the old value. This is safe due to the fact that when the `Adds` entry is created, the `Old` value is stored before adding the `Adds` entry, so for the first `Adds` entry the `Old` value does not yet exist and will be added.

This removes extraneous `Old` entries that did not actually originally exist. Before this change `ChangeState.Revert` could have restored entries the should not exists, based on these extraneous `Old` entries.

Note that the change in `TestMapState_AccumulateMapChangesOnVisibilityKeys` on the 3rd commit shows the effect of this change. Previously the test resulted in an entry in `ChangeState.Old` that did not exist before the changes started, but was recoded due to an entry being first added and then modified within the same set of changes.

> Note for backporting: Releases prior to 1.14 should only backport the last commit, as the changes addressed in first two are only in the main branch and are included to make the backports of the last commit easier.

```release-note
Policy revert used in rare error cases has been corrected.
```

### CODE REVIEW

File: pkg/policy/mapstate.go [hunk 1]
```diff
@@ -65,6 +65,7 @@ type MapState interface {
 	AddVisibilityKeys(PolicyOwner, uint16, *VisibilityMetadata, ChangeState)
 	Len() int
 	Equals(MapState) bool
+	Diff(expected MapState) string
```
Comment (bimmlerd) [inline pkg/policy/mapstate.go]: suggestion: If it's just for use in testing, can we explicitly take a `testing.T` argument (even if we don't need it) to ensure it's not used elsewhere?

**Review by bimmlerd — COMMENTED** [review]
drive-by comment - lgtm otherwise

File: pkg/policy/mapstate.go (hunk 1 above)
Comment (jrajahalme) [reply pkg/policy/mapstate.go]: good point, will do!

Comment (jrajahalme) [reply pkg/policy/mapstate.go]: Added!

**Review by nebril — APPROVED** [review]

**Review by bimmlerd — APPROVED** [review]

**Review by aditighag — APPROVED** [review]
Approving `pkg/endpoint/bpf.go` changes for my codeowners.

**Comment by aanm** [comment]
This PR is already part of 1.15

**Comment by jrajahalme** [comment]
Looked into backporting this to 1.13 and 1.12, and there are a lot of non-backported dependencies that make the backport somewhat risky. Since we don't know of any reports of actual problems cause by the issue fixed here, I'll drop the 1.13 and 1.12 backports for now.

**Pull request merged**